pkg/manifests/nginx.go

```diff
@@ -560,8 +560,8 @@ func newNginxIngressControllerHPA(conf *config.Config, ingressConfig *NginxIngre
 				Name:       ingressConfig.ResourceName,
 			},
 			MinReplicas:                    util.Int32Ptr(2),
-			MaxReplicas:                    100,
-			TargetCPUUtilizationPercentage: util.Int32Ptr(80),
+			MaxReplicas:                    500,
+			TargetCPUUtilizationPercentage: util.Int32Ptr(70),
```

Comment:
> In short, I concluded that this was due to nginx waiting too long to scale and overcompensating by creating too many replicas too quickly, which suddenly had to be drained and removed when the load evened out/was no longer ramping up, which led to worse performance in our tests.


this sentence in your description seems to imply the opposite of what you're doing here.

Reply:
So I did try dropping to 50 as well.
On one hand, that resulted in worse performance for our RPS tests. I'm not really sure why to be completely honest, but our rate of 504s was significantly higher - and increasing server replicas with this lower threshold did nothing to change that. 

However, it performed better for our larger request tests - our error rate went down and we experienced less socket issues, which may be due to just having more replicas to accept more connections as the number of rps grew rapidly.

I think this highlights the point that the scaling requirements that customers will have will be dependent upon a variety of factors, including the kind of traffic that they expect/serve. 

That said, I think that 1MB request and response bodies (which performed better with this lower CPU threshold) is a bit larger than the individual responses that most applications will be transmitting, so I'd rather cater towards the smaller request size until customers can make that decision for themselves.

Reply:
https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#configurable-scaling-behavior
specifically, scaling policies are an alternative to what is proposed.

Reply:
I want to see prometheus metrics for cpu over time for these tests. Not throwing a dart at a board and seeing what sticks

Reply:
Regarding the description, I'll flesh the logic out a bit more.

Suppose we have a set of nginx replicas, and let's say they're experiencing a growing request load. Let's also call the amount of time between a replica hitting 70% CPU usage and 80% cpu usage t. By the time that new replicas have hit 80%, the rest of the replicas have been getting overwhelmed for t longer. Until new replicas can kick in and receive new requests, the CPU usage observed on these active replicas remains at a value greater than the threshold for longer and longer, which causes the hpa to continue to churn out more replicas.

If, instead, we scale at 70%, that decreases the amount of time before new replicas are spawned and can begin accepting requests, and resultantly the amount of time that our active replicas are dangerously close to maxing out their CPU.

There's another effect at play too which I've observed, and I'll describe that below.

In the first situation, let's say once traffic has stopped increasing, the added replicas become active. However, enough new replicas have been created that the average CPU usage has actually dropped below our goal - at which point the HPA will begin removing replicas. During periods of replica removal, our load test clients saw an increase in socket issues, but once the number of replicas stabilized, those went to zero.

In the second, if new replicas are able to join the cluster more quickly during an increasing request load, that reduces the amount of time between our existing replicas being overwhelmed and new ones being able to join the deployment, right? At which point, the new replicas can begin shouldering the burden of new requests sooner, which results in the previously-added-but-ramping-up replicas not needing to cross the CPU threshold as quickly, which results in less of an "overshot" in extra replicas. In these cases with the lower threshold, the dropoff in replicas that took place after rps stabilized was much lower than for tests with the higher CPU threshold, and as a result, the overall percentage of requests of socket issues we saw was lower.

Now, I still don't know why the RPS/small request tests struggled when the CPU was too low (50%). If I had to guess, it might have something to do with how nginx scales when there are too many replicas that aren't being used enough. Regardless, that behavior was demonstrated over multiple iterations in our testing, where the error rate was over twice as high as tests with a 70% threshold. Hence the in-between value of 70.

Reply:
why change threshhold vs cpu request?